packaging/man/coolercontrol.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.3.
.TH COOLERCONTROL "1" "Oktober 2025" "CoolerControl 3.0.2" "User Commands"
.SH NAME
CoolerControl \- manual page for CoolerControl 3.0.2
.SH SYNOPSIS
.B coolercontrol
[\fI\,options\/\fR]
.SH DESCRIPTION
CoolerControl GUI Desktop Application
.SH OPTIONS
.TP
\fB\-h\fR, \fB\-\-help\fR
Displays help on commandline options.
.TP
\fB\-\-help\-all\fR
Displays help, including generic Qt options.
.TP
\fB\-v\fR, \fB\-\-version\fR
Displays version information.
.TP
\fB\-d\fR, \fB\-\-debug\fR
Enable debug output.
.TP
\fB\-\-full\-debug\fR
Enable full debug output. This outputs a lot of data.
.TP
\fB\-\-disable\-gpu\fR
Disable GPU hardware acceleration.
.SH "SEE ALSO"
The full documentation for
.B CoolerControl
is maintained as a Texinfo manual.  If the
.B info
and
.B CoolerControl
programs are properly installed at your site, the command
.IP
.B info CoolerControl
.PP
should give you access to the complete manual.


CHANGELOG.md
# Changelog

<!-- trunk-ignore-all(markdownlint/MD024): Multiple heading with same content -->

<!--
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
Release notes are automatically generated from this file and git tags.
-->

## [3.0.2] - 2025-11-03

### Added

- firmware profile support for various HWMon drivers (!366)
- simulated zero-rpm mode for Nvidia GPUs that don't support 0% manual fan speed (#523)

### Changed

- minor liqctld improvements (!363)
- improved downstream packaging support (!364, !365, !369)
- improved UI reconnection logic (!368, #520)

### Fixed

- legacy Asetec/Kraken device support (#515)
- liqctld shutdown issue (!367)

## [3.0.1] - 2025-10-04

### Changed

- improved multiple-UI reconnection handling

### Fixed

- liqctld service failure affecting some liquidctl devices
- RPM warning about unversioned obsoletes

## [3.0.0] - 2025-10-03

### Removed

- The `coolercontrol-liqctld` package and service has been removed. The service has now been
  embedded in the `coolercontrold` daemon as a child process (!340)

### Added

- Overlay Profiles: Advanced offset controls on top of base Profiles (!347)
- Custom Sensor Offset and parent-child chaining (!346)
- Custom Profile temperature ranges (!344)
- Ability to import and export color schemes
- API changes to help 3rd party integrations
- Keyboard shortcuts for common navigation (!349)
- New device/main menu: drag and drop device and sensor sorting, pinned sensors, host name,
  additional submenu options, color control for device and entity titles, and improved usability and
  performance (!339)
- Firmware-controlled profile support for some AMDGPUs and liquidctl devices. (hwmon device support
  coming soon) (!354)
- Foundation for more advanced device and channel specific controls
- Min and Max CPU core frequency metrics
- Mix Profile difference mix function (!348)
- Ability to change liquidctl log level with environment variable

### Changed

- New Color Picker UI component
- Various logging improvements
- the `liqctld` service logs are now unified with the daemon's normal log output
- Nearly all dependencies upgraded, including several major version changes
- Website and documentation improvements
- Improved metric standards in the UI (!345)
- UI Performance and reloading improvements

### Fixed

- Issue with 24h time format
- Issue with minimum fan duty threshold (!338)
- Various minor bugs and UI issues

## [2.2.2] - 2025-07-18

### Added

- Support for devices that only have non-controllable fans (#465)
- aarch64/arm64 & musl build support (!327, #475)
- Support for raspberry cpus (!329)
- DockerHub builds (!331)
- Alert warmup duration (#466)
- Show all sensors with chart toggle (#480)
- Show running instance when starting desktop app (#453)

### Changed

- Updated dependencies (!336)

### Fixed

- liquidctl Coolit driver support (#467)
- Hover menu not displaying correctly (#468)
- ThinkPad full-speed mode not disengaging (#470)
- Daemon status translation key (!328)
- Duplicate welcome text (#477)
- Minor Alerts related issues (!333)

## [2.2.1] - 2025-06-13

### Added

- Improved security and quality checks for the GitLab CI Pipeline (!312)

### Changed

- Improved test coverage for main control engine (!315)
- Updated copyright year (!316)

### Fixed

- Disabled support for AsusRyujin liquidctl driver (#457)
- Removed no longer used title (!318)

### Security

- Vulnerability issue with dependency (!314)

## [2.2.0] - 2025-05-31

### Added

- Support for older Intel CPUs (#437)
- Support for liquidctl MPGCooler driver (#436)
- Support for liquidctl AsusRyujin driver (#441)
- Support for liquidctl Coolit Hydro GT driver (#442)
- Zero Temperature Drive Power State (#443)
- Project Vision document (!297)
- International Language support i18n (!296 @kuilei0926)

coolercontrol-gui/coolercontrol/coolercontrol.py
#  CoolerControl - monitor and control your cooling and other devices
#  Copyright (c) 2022  Guy Boldon
#  |
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#  |
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#  |
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <https://www.gnu.org/licenses/>.
# ----------------------------------------------------------------------------------------------------------------------

import argparse
import importlib.metadata
import logging.config
import os
import platform
import sys
import textwrap
import time
import traceback
from logging.handlers import RotatingFileHandler

import setproctitle
from PySide6 import QtCore
from PySide6.QtCore import QTimer, QCoreApplication, QEvent, QSize, QPoint
from PySide6.QtGui import QColor, Qt, QIcon, QAction, QShortcut, QKeySequence, QHideEvent, QShowEvent
from PySide6.QtWidgets import QMainWindow, QGraphicsDropShadowEffect, QApplication, QSystemTrayIcon, QMenu, QHBoxLayout

from coolercontrol.app_instance import ApplicationInstance
from coolercontrol.dialogs.connection_failure_dialog import ConnectionFailureDialog
from coolercontrol.exceptions.daemon_connection_error import DaemonConnectionError
from coolercontrol.exceptions.restart_needed import RestartNeeded
from coolercontrol.services.app_updater import AppUpdater
from coolercontrol.services.dynamic_buttons import DynamicButtons
from coolercontrol.settings import Settings, UserSettings, IS_APP_IMAGE, FeatureToggle
from coolercontrol.view.core.functions import Functions
from coolercontrol.view.uis.pages.info_page import InfoPage
from coolercontrol.view.uis.pages.settings_page import SettingsPage
from coolercontrol.view.uis.windows.main_window import SetupMainWindow, UI_MainWindow, MainFunctions
from coolercontrol.view.uis.windows.splash_screen.splash_screen_style import SPLASH_SCREEN_STYLE
from coolercontrol.view.uis.windows.splash_screen.ui_splash_screen import Ui_SplashScreen  # type: ignore
from coolercontrol.view.widgets.py_frameless_window import FramelessMainWindow, StandardTitleBar, FramelessDialog
from coolercontrol.view_models.devices_view_model import DevicesViewModel

logging.config.fileConfig(Settings.app_path.joinpath('config/logging.conf'), disable_existing_loggers=False)
log = logging.getLogger(__name__)
_APP: QApplication
_INIT_WINDOW: QMainWindow
_ICON: QIcon
_RUNNING_INSTANCE: ApplicationInstance


class Initialize(FramelessMainWindow):

    def __init__(self) -> None:
        super().__init__()
        log.info("CoolerControl is initializing...")
        self._load_progress_counter: int = 0
        self.setResizeEnabled(False)

        self.app_settings = Settings.app
        self.user_settings = Settings.user
        self.user_settings.setValue('version', self.app_settings['version'])
        self.theme = Settings.theme
        QApplication.setApplicationName(self.app_settings['app_name'])
        QApplication.setApplicationVersion(self.app_settings['version'])

        _verify_single_running_instance()

        # Setup splash window
        self.ui = Ui_SplashScreen()
        self.ui.setupUi(self)
        self.ui.centralwidget.setStyleSheet(f"""
            #centralwidget {{
                background-color: {self.theme["app_color"]["bg_one"]};
                border-radius: 14;
                border: 2px solid {self.theme["app_color"]["dark_one"]};
            }}
        """)
        self.ui.verticalLayout.setContentsMargins(2, 2, 2, 2)
        splash_style = SPLASH_SCREEN_STYLE.format(
            _bg_color=self.theme["app_color"]["bg_one"],
            _title_color=self.theme["app_color"]["text_title"],
            _color=self.theme["app_color"]["text_foreground"],
            _progress_bg_color=self.theme["app_color"]["bg_two"],
            _progress_color=self.theme["app_color"]["white"],
            _progress_from_color=self.theme["app_color"]["icon_hover"],
            _progress_to_color=self.theme["app_color"]["context_pressed"]
        )
        self.ui.dropShadowFrame.setStyleSheet(splash_style)
        self.ui.label_title.setStyleSheet(splash_style)
        self.ui.label_description.setStyleSheet(splash_style)
        self.ui.label_loading.setStyleSheet(splash_style)
        self.ui.label_version.setStyleSheet(splash_style)
        self.ui.progressBar.setStyleSheet(splash_style)
        self.setAttribute(Qt.WA_TranslucentBackground)
        # Shadows are not working well anymore at all:
        # self.shadow = QGraphicsDropShadowEffect(self)
        # self.shadow.setBlurRadius(20)
        # self.shadow.setXOffset(0)
        # self.shadow.setYOffset(0)
        # self.shadow.setColor(QColor(0, 0, 0, 60))
        # self.ui.dropShadowFrame.setGraphicsEffect(self.shadow)
        self.titleBar.minBtn.hide()
        self.titleBar.maxBtn.hide()
        self.titleBar.closeBtn.hide()
        self.titleBar.setDoubleClickEnabled(False)
        self.titleBar.raise_()

        self.ui.label_loading.setText("<strong>Initializing</strong>")
        self.ui.label_version.setText(f'<strong>version</strong>: {self.app_settings["version"]}')

        if Settings.user.value(UserSettings.START_MINIMIZED, defaultValue=False, type=bool):
            if not Settings.user.value(UserSettings.HIDE_ON_MINIMIZE, defaultValue=False, type=bool):
                self.showMinimized()
        else:
            self.show()

        self.timer = QTimer()
        self.timer.timeout.connect(self.init_devices)
        self.timer.start(10)

    def init_main_window(self):
        self.main = MainWindow()
        self.main.devices_view_model = DevicesViewModel()
        self.main.dynamic_buttons = DynamicButtons(
            self.main.devices_view_model,
            self.main
        )

    def init_devices(self) -> None:
        try:
            _APP.setQuitOnLastWindowClosed(False)  # splash and dialog windows at startup
            should_check_for_update: bool = self.user_settings.value(
                UserSettings.CHECK_FOR_UPDATES, defaultValue=False, type=bool
            ) and IS_APP_IMAGE and FeatureToggle.appimage_update
            if self._load_progress_counter == 0:
                self.ui.label_loading.setText("<strong>Startup</strong> delay")
            elif self._load_progress_counter == 2:
                time.sleep(0.5)  # we initialize the main window after a delay to help with system tray issues
                self.init_main_window()

            elif self._load_progress_counter == 5:

                if should_check_for_update:
                    self.ui.label_loading.setText("<strong>Checking</strong> for updates")
            elif self._load_progress_counter == 10:
                if should_check_for_update:
                    AppUpdater.run(self)

                self.ui.label_loading.setText("<strong>Connecting</strong> to daemon")
            elif self._load_progress_counter == 50:
                try:
                    self.main.devices_view_model.init_devices_from_daemon()
                    if self.main.devices_view_model.devices:
                        self.main.devices_view_model.schedule_status_updates()
                except RestartNeeded:
                    log.info("Restart Required, shutting down...")
                    self.force_shutdown()
                except DaemonConnectionError:
                    ConnectionFailureDialog().display()
                    self.force_shutdown()

                self.ui.label_loading.setText("<strong>Initializing</strong> the UI")
            elif self._load_progress_counter == 75:
                # finalize repo setup
                self.main.devices_view_model.init_scheduler_commander()
                # wire up core logic:
                self.main.devices_view_model.subscribe(self.main.ui.system_overview_canvas)
                self.main.dynamic_buttons.create_menu_buttons_from_devices()
                self.main.ui.left_column.menus.info_page_layout.addWidget(
                    InfoPage(self.main.devices_view_model.devices)
                )
                self.main.ui.left_column.menus.settings_page_layout.addWidget(SettingsPage())

            elif self._load_progress_counter >= 100:
                self.timer.stop()
                log.info("Displaying Main UI Window...")
                if Settings.user.value(UserSettings.START_MINIMIZED, defaultValue=False, type=bool):
                    if Settings.user.value(UserSettings.HIDE_ON_MINIMIZE, defaultValue=False, type=bool):
                        self.main.ui.system_overview_canvas.animation.pause()  # pause animations at startup if hidden
                    else:
                        self.main.showMinimized()
                else:
                    self.main.show()
                self.close()

            self._load_progress_counter += 1
            self.ui.progressBar.setValue(self._load_progress_counter)
        except BaseException as ex:
            log.fatal('Unexpected Error', exc_info=ex)
            self.force_shutdown()

    def force_shutdown(self) -> None:
        log.info("Shutting down...")
        _APP.setQuitOnLastWindowClosed(True)
        self.main.devices_view_model.shutdown()
        self.close()


class CustomTitleBar(StandardTitleBar):
    """ Custom title bar """

    def __init__(self, parent, icon_w_h: int = 20):
        super().__init__(parent, icon_w_h)

        self.titleLabel.setStyleSheet(
            f'font: {Settings.app["font"]["title_size"]}pt "{Settings.app["font"]["family"]}"; '
            f'color: {Settings.theme["app_color"]["text_title"]};'
            'padding: 0 4px;'
            'background: transparent;'
        )
        self.minBtn.setNormalColor(Settings.theme["app_color"]["icon_color"])
        self.minBtn.setHoverColor(Settings.theme["app_color"]["icon_hover"])
        self.minBtn.setHoverBackgroundColor(Settings.theme["app_color"]["bg_three"])
        self.minBtn.setPressedColor(Settings.theme["app_color"]["icon_pressed"])

        self.maxBtn.setNormalColor(Settings.theme["app_color"]["icon_color"])
        self.maxBtn.setHoverColor(Settings.theme["app_color"]["icon_hover"])
        self.maxBtn.setHoverBackgroundColor(Settings.theme["app_color"]["bg_three"])
        self.maxBtn.setPressedColor(Settings.theme["app_color"]["icon_pressed"])

        self.closeBtn.setNormalColor(Settings.theme["app_color"]["icon_color"])
        self.closeBtn.setHoverColor(Settings.theme["app_color"]["icon_hover"])
        self.closeBtn.setHoverBackgroundColor(Settings.theme["app_color"]["red"])
        self.closeBtn.setPressedColor(Settings.theme["app_color"]["icon_pressed"])


class MainWindow(FramelessMainWindow):

    def __init__(self) -> None:
        super().__init__()
        sys.excepthook = self.log_uncaught_exception
        self.app_settings = Settings.app
        self.user_settings = Settings.user
        self.theme = Settings.theme
        self.ui = UI_MainWindow()
        self.setTitleBar(CustomTitleBar(self, icon_w_h=28))
        self.ui.title_bar = self.titleBar
        self.ui.setup_ui(self)
        self.setWindowTitle(self.app_settings["app_name"])
        self.setWindowIcon(QIcon(Functions.set_svg_image("logo_color.svg")))
        self.active_left_sub_menu: str = ''
        self.devices_view_model: DevicesViewModel = None  # type: ignore
        self.dynamic_buttons: DynamicButtons = None  # type: ignore
        SetupMainWindow.setup_gui(self)

        # restore window size & position
        if self.user_settings.contains(UserSettings.WINDOW_SIZE):
            try:
                self.resize(  # type: ignore
                    self.user_settings.value(
                        UserSettings.WINDOW_SIZE,
                        defaultValue=QSize(self.app_settings["startup_size"][0], self.app_settings["startup_size"][1]),
                        type=QSize
                    )
                )
                self.move(  # type: ignore
                    self.user_settings.value(
                        UserSettings.WINDOW_POSITION,
                        defaultValue=QPoint(200, 200),
                        type=QPoint
                    )
                )
                log.debug('Loaded saved window size')
            except BaseException as ex:
                log.error('Unable to get and restore saved window geometry: %s', ex)

        tray_icon_style = 'white' \
            if Settings.user.value(UserSettings.ENABLE_BRIGHT_TRAY_ICON, defaultValue=False, type=bool) \
            else 'color'
        tray_icon = QIcon(Functions.set_svg_image(f'logo_{tray_icon_style}.svg'))
        tray_icon.setIsMask(True)
        self.tray_menu = QMenu(self)
        self.tray_menu.addAction(
            QAction(
                self.app_settings['app_name'], self, icon=QIcon(tray_icon), triggered=None, enabled=False
            )  # type: ignore[call-overload]
        )
        self.tray_menu.addSeparator()
        self.tray_menu.addAction(  # shortcut='Ctrl+h' - shortcuts don't appear to work for the sys tray actions?
            QAction('&Show', self, triggered=self.show_main_window))  # type: ignore[call-overload]
        self.tray_menu.addAction(
            QAction('&Quit', self, triggered=self.force_close))  # type: ignore[call-overload]
        self.tray = QSystemTrayIcon(self)
        self.tray.setIcon(tray_icon)
        self.tray.setVisible(True)
        self.tray.setContextMenu(self.tray_menu)

        self.shortcut_close = QShortcut(QKeySequence('Ctrl+Q'), self)
        self.shortcut_close.activated.connect(self.force_close)
        self.shortcut_hide = QShortcut(QKeySequence('Ctrl+H'), self)
        self.shortcut_hide.activated.connect(self.hide)
        self.shortcut_toggle_menu = QShortcut(QKeySequence('Ctrl+/'), self)
        self.shortcut_toggle_menu.activated.connect(self.ui.left_menu.toggle_animation)

    def show_main_window(self) -> None:
        if not self.isVisible():
            self.setVisible(True)
        self.activateWindow()

    # main menu btn
    def btn_clicked(self) -> None:

        btn = SetupMainWindow.setup_btns(self)
        btn_id = btn.objectName()
        log.debug('Button %s, clicked!', btn_id)

        # home btn
        if btn_id == "btn_system":
            self.ui.left_menu.select_only_one(btn.objectName())
            self.clear_left_sub_menu()
            MainFunctions.set_page(self, self.ui.load_pages.system_overview)
            self.ui.system_overview_canvas.redraw_workaround()

        # Info and Settings combined:
        elif btn_id in ["btn_settings", "btn_info"]:
            if not MainFunctions.left_column_is_visible(self):
                self.dynamic_buttons.uncheck_all_channel_buttons()
                MainFunctions.toggle_left_column(self)
                self.ui.left_menu.select_only_one_tab(btn_id)
                self.active_left_sub_menu = btn_id
            elif btn_id == self.active_left_sub_menu:
                # close side menu
                self.ui.left_menu.deselect_all_tab()
                self.active_left_sub_menu = ''
                MainFunctions.toggle_left_column(self)
            else:
                self.active_left_sub_menu = btn.objectName()
                self.ui.left_menu.select_only_one_tab(btn.objectName())

            if btn_id == "btn_settings":
                MainFunctions.set_left_column_menu(
                    self,
                    menu=self.ui.left_column.menus.settings_page,
                    title="Settings",
                    icon_path=Functions.set_svg_icon("icon_settings.svg")
                )
            elif btn_id == "btn_info":
                MainFunctions.set_left_column_menu(
                    self,
                    menu=self.ui.left_column.menus.info_page,
                    title="Info",
                    icon_path=Functions.set_svg_icon("icon_info.svg")
                )
        else:
            self.dynamic_buttons.set_device_page(btn_id)

    def clear_left_sub_menu(self) -> None:
        self.ui.left_menu.deselect_all_tab()
        if MainFunctions.left_column_is_visible(self):
            MainFunctions.toggle_left_column(self)

    def btn_released(self) -> None:
        btn = SetupMainWindow.setup_btns(self)
        log.debug('Button %s, released!', btn.objectName())

    def hideEvent(self, event: QHideEvent) -> None:
        """improved efficiency by pausing animations & line calculations when window is hidden"""
        self.ui.system_overview_canvas.animation.pause()
        if MainFunctions.device_column_is_visible(self):
            MainFunctions.toggle_device_column(self)
        self.dynamic_buttons.uncheck_all_channel_buttons()

    def showEvent(self, event: QShowEvent) -> None:
        if self.ui.system_overview_canvas.animation.event_source:
            self.ui.system_overview_canvas.animation.event_source.interval = 100
        self.ui.system_overview_canvas.animation.resume()

    def resizeEvent(self, event: QEvent) -> None:
        if self.ui.device_column_frame.width() > 0:
            self.ui.device_column_frame.setMinimumWidth(int((self.width() - self.ui.left_menu_frame.width()) / 2))

    def closeEvent(self, event: QEvent) -> None:
        """Shutdown or minimize to tray"""
        _APP.setQuitOnLastWindowClosed(True)  # set the safe default again, just in case of an unexpected crash
        if self.user_settings.value(UserSettings.HIDE_ON_CLOSE, defaultValue=False, type=bool):
            self.hide()
            event.ignore()
        else:
            self.shutdown(event)

    def force_close(self) -> None:
        if self.user_settings.value(UserSettings.HIDE_ON_CLOSE, defaultValue=False, type=bool):
            self.shutdown()
        else:
            self.close()

    def shutdown(self, event: QEvent | None = None) -> None:
        """Shutdown process"""
        log.info("Shutting down...")
        self.devices_view_model.shutdown()
        if self.user_settings.value(UserSettings.SAVE_WINDOW_SIZE, defaultValue=True, type=bool):
            if not self.isMaximized():  # do not save maximized size
                self.user_settings.setValue(UserSettings.WINDOW_SIZE, self.size())
                self.user_settings.setValue(UserSettings.WINDOW_POSITION, self.pos())
                log.debug('Saved window size in user settings')
        else:
            self.user_settings.remove(UserSettings.WINDOW_SIZE)
            self.user_settings.remove(UserSettings.WINDOW_POSITION)
        self.close()
        _APP.quit()

    @staticmethod
    def log_uncaught_exception(*exc_info: tuple) -> None:
        text = "".join(traceback.format_exception(*exc_info))
        log.error('Unexpected error has occurred: %s', text)


def _verify_single_running_instance() -> None:
    global _RUNNING_INSTANCE
    _RUNNING_INSTANCE = ApplicationInstance()


def _get_version_attribute(package_object: object) -> str:
    return getattr(package_object, "__version__", "unknown")


def _get_package_version(package_name: str) -> str:
    """This searches for package versions.
    First it checks the metadata, which is present for all packages.
    If the metadata isn't found, like with the compiled AppImage, it checks inside the package for __version__.
    If package doesn't exist then it either defaults to the last known version or "unknown"
    """
    try:
        return importlib.metadata.version(package_name)
    except importlib.metadata.PackageNotFoundError:
        match package_name:
            case "pillow":
                import PIL
                return _get_version_attribute(PIL)
            case "pyside6":
                import PySide6
                return _get_version_attribute(PySide6)
            case "matplotlib":
                import matplotlib
                return _get_version_attribute(matplotlib)
            case "numpy":
                import numpy
                return _get_version_attribute(numpy)
            case _:
                return "unknown"


def _system_info() -> str:
    sys_info: str = textwrap.dedent(f'''
    {Settings.app["app_name"]} {Settings.app["version"]}

    System:''')
    if platform.system() == 'Linux':
        sys_info += f'\n    {platform.freedesktop_os_release().get("PRETTY_NAME")}'  # type: ignore
    pyside_version: str = _get_package_version("pyside6_essentials")
    if pyside_version == "unknown":
        pyside_version = _get_package_version("pyside6")
    sys_info += textwrap.dedent(f'''
        {platform.platform()}

    Dependency versions:
        Python     {platform.python_version()}
        Pillow     {_get_package_version("pillow")}
        PySide6    {pyside_version}
        Matplotlib {_get_package_version("matplotlib")}
        Numpy      {_get_package_version("numpy")}
    ''')
    return sys_info


def _init_cli_and_logging() -> None:
    parser = argparse.ArgumentParser(
        description='monitor and control your cooling and other devices',
        exit_on_error=False,
        formatter_class=argparse.RawTextHelpFormatter
    )
    parser.add_argument(
        '-v', '--version', action='version',
        version=f'\n {_system_info()}'
    )
    parser.add_argument("--debug", action="store_true",
                        help="enable debug output\n"
                             "a log file is created under /tmp/coolercontrol/")
    args = parser.parse_args()
    if args.debug:
        log_filename = Settings.tmp_path.joinpath('coolercontrol.log')
        file_handler = RotatingFileHandler(
            filename=log_filename, maxBytes=10485760, backupCount=5, encoding='utf-8'
        )
        log_formatter = logging.getLogger('root').handlers[0].formatter
        file_handler.setFormatter(log_formatter)
        logging.getLogger('root').setLevel(logging.DEBUG)
        logging.getLogger('root').addHandler(file_handler)
        logging.getLogger('matplotlib').setLevel(logging.INFO)
        logging.getLogger('matplotlib').addHandler(file_handler)
        logging.getLogger('apscheduler').setLevel(logging.INFO)
        logging.getLogger('apscheduler').addHandler(file_handler)
        logging.getLogger('liquidctl').setLevel(logging.DEBUG)
        logging.getLogger('liquidctl').addHandler(file_handler)
        log.debug('DEBUG level enabled\n%s', _system_info())


def main() -> None:
    setproctitle.setproctitle("coolercontrol")
    _init_cli_and_logging()
    if os.geteuid() == 0:
        log.error("This GUI program should not be run as root.")
        sys.exit(1)
    QCoreApplication.setAttribute(QtCore.Qt.AA_ShareOpenGLContexts)
    QApplication.setAttribute(QtCore.Qt.AA_EnableHighDpiScaling)
    QApplication.setAttribute(QtCore.Qt.AA_UseHighDpiPixmaps)
    QApplication.setAttribute(Qt.AA_UseDesktopOpenGL)
    QApplication.setAttribute(Qt.AA_Use96Dpi)
    os.environ['QT_FONT_DPI'] = '96'  # this appears to need to be set to keep things sane
    os.environ['QT_SCALE_FACTOR'] = str(  # scale performs better than higher dpi
        Settings.user.value(UserSettings.UI_SCALE_FACTOR, defaultValue=1.0, type=float)
    )
    if Settings.app["custom_title_bar"]:
        os.environ['QT_WAYLAND_DISABLE_WINDOWDECORATION'] = '1'  # to make sure wayland doesn't do system decorations
    global _APP, _ICON, _INIT_WINDOW
    _APP = QApplication(sys.argv)
    _ICON = QIcon(Functions.set_svg_image('logo_color.svg'))
    _APP.setWindowIcon(_ICON)
    _INIT_WINDOW = Initialize()
    sys.exit(_APP.exec())


if __name__ == "__main__":
    main()
